cmd/monaco/dynatrace/dynatrace.go

```diff
@@ -33,92 +30,87 @@ import (
-		return canEstablishClassicConnection(ctx, env)
+	classicUrl := env.URL.Value
+
+	// check if the OAuth connection works and get the classicURL in order to check the token authentication next if given
+	if env.Auth.OAuth != nil {
+		oauthCreds := clientcredentials.Config{
+			ClientID:     env.Auth.OAuth.ClientID.Value.Value(),
+			ClientSecret: env.Auth.OAuth.ClientSecret.Value.Value(),
+			TokenURL:     env.Auth.OAuth.GetTokenEndpointValue(),
+		}
+		var err error
+		if classicUrl, err = getDynatraceClassicURL(ctx, env.URL.Value, oauthCreds); err != nil {
+			err = formatAuthError(env, env.URL.Value, err, "OAuth")
+			return fmt.Errorf("please verify that this environment is a Dynatrace Platform environment. %w", err)
+		}
 	}
 
-	return isPlatformEnvironment(ctx, env)
+	if env.Auth.Token != nil {
+		if err := validateEstablishClassicConnection(ctx, env, classicUrl); err != nil {
+			return err
+		}
+	}
+	return nil
 }
 
-// canEstablishClassicConnection checks if a classic connection (via token) can be established. Scopes are not validated.
-func canEstablishClassicConnection(ctx context.Context, env manifest.EnvironmentDefinition) bool {
+// validateEstablishClassicConnection checks if a classic connection (via token) can be established. Scopes are not validated.
+func validateEstablishClassicConnection(ctx context.Context, env manifest.EnvironmentDefinition, classicURL string) error {
 	token := env.Auth.Token.Value.Value()
 	client, err := clients.Factory().
-		WithClassicURL(env.URL.Value).
+		WithClassicURL(classicURL).
 		WithAccessToken(token).
 		WithRateLimiter(true).
 		WithRetryOptions(&client.DefaultRetryOptions).
 		CreateClassicClient()
 	if err != nil {
-		report.GetReporterFromContextOrDiscard(ctx).ReportLoading(report.StateError, fmt.Errorf("could not create client %q (%s): %w", env.Name, env.URL.Value, err), "", nil)
-		log.Error("Could not create client %q (%s): %v", env.Name, env.URL.Value, err)
-		return false
+		return fmt.Errorf("could not create client %q (%s): %w", env.Name, classicURL, err)
 	}
 
 	if _, err := apitoken.GetTokenMetadata(ctx, client, token); err != nil {
-		handleAuthError(ctx, env, err)
-		log.Error("Please verify that this environment is a Dynatrace Classic environment.")
-		return false
-	}
-	return true
-}
-
-func isPlatformEnvironment(ctx context.Context, env manifest.EnvironmentDefinition) bool {
-	oauthCreds := clientcredentials.Config{
-		ClientID:     env.Auth.OAuth.ClientID.Value.Value(),
-		ClientSecret: env.Auth.OAuth.ClientSecret.Value.Value(),
-		TokenURL:     env.Auth.OAuth.GetTokenEndpointValue(),
-	}
-
-	if _, err := getDynatraceClassicURL(ctx, env.URL.Value, oauthCreds); err != nil {
-		handleAuthError(ctx, env, err)
-		log.Error("Please verify that this environment is a Dynatrace Platform environment.")
-		return false
+		err = formatAuthError(env, classicURL, err, "token")
+		return fmt.Errorf("please verify that this environment is a Dynatrace Classic environment. %w", err)
```

Comment:
I think "please verify that this environment is a Dynatrace Classic environment" belongs in a log message but not an error message. I would move this out of this function.

Reply:
In fact, the call to `formatAuthError(...)` should also move out, at least to line 81 above, or, both calls in _that_ function could move out. 

In a way, I see `formatAuthError(...)` as a code smell that suggests we should probably just wrap any error returned from `VerifyEnvironmentAuthentication(...)` with the information there rather than calling this twice inside.